sidebars: Enable up and down arrow keys in user and stream search.

florianjuengermann opened the pull request:
This fixes [#5920](https://github.com/zulip/zulip/issues/5920).

As soon as the search-bar is in focus, the first element becomes highlighted by an assigned CSS class. Then, the key-down events are processed to find arrow key presses and change the element in focus.

The tests are done with Casper as the user interactions can be implemented more easily there compared to the Node tests.

![image](https://user-images.githubusercontent.com/13962498/33527659-caa93522-d854-11e7-9983-675be5429d2b.png)

showell commented:
@rishig @brockwhittaker I'm punting this back to you guys to review first, from a usability perspective and also opinions on how this would work code-wise with list_render.

@florianjuengermann Thanks for working on this!  I really want this feature.

rishig commented:
Assigned to Brock, thanks!

brockwhittaker commented:
Couple issues:

1. There's too much code duplication. You add a few functions like `keydown_stream_filter` mulitiple times to the code base. Perhaps find a way to make it a generic.

2. In the user sidebar if it is long you can just keep down arrow-ing off screen and it doesn't scroll with it.

Otherwise seems to look good!

florianjuengermann commented:
@brockwhittaker Thanks for your feedback. Finally, I found some time to implement it. I created generic functions in `stream_list.js` and then used them for both sidebars. I know that the location is not ideal but I did not want to move it `ui.js` because this file is already packed.
For the automatical scrolling, there was a function I could use without problems.
I also reorganized the commits so that they are independent for one another.

florianjuengermann commented:
@brockwhittaker I have updated the pull request to adapt to the recent commit concerning the right sidebar.

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+        exports.clear_and_hide_search();
+        narrow.by('stream', top_stream.name,
+            {select_first_unread: true, trigger: 'sidebar enter key'});
+    }
```
If there's no stream, should the behavior here still capture the `enter` keypress?  Just wondering if this function should return `handled` to the caller.

florianjuengermann replied:
Normally, the ```enter``` press needs to be suppressed to avoid a line-break in the compose box. If there is no stream, the suppression of the ```enter``` press guarantees that no line-break character is passed to the ```<input>```. I think the ```<input>``` cannot receive a line-break character, so the suppression is probably redundant but I would leave it like that.
I added a clarifying comment in line 653.

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+        case 38: {
+            // Up-arrow key was pressed
+            keydown_arrow_key(true, all_streams_selector,
+                             scroll_container, highlighting_class);
```
These are missing a space of indentation; I wonder why the linter didn't catch this?

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+
+
+exports.keydown_filter = function (e, all_streams_selector,
+ scroll_container, highlighting_class, enter_press_function) {
```
This second line should be indented consistent with the rest of the project.

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+}
+
+function keydown_arrow_key(dir_up, all_streams_selector,
+                         scroll_container, highlighting_class) {
```
This should be a couple spaces more indented.

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+
+        var current_sel = $(all_streams_selector + '.' + highlighting_class).expectOne();
+
+        var next_sibling = next_sibing_in_dir(current_sel, dir_up);
```
I think the two blank lines above are unnecessary for readability

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+    // Is there at least one stream?
+    if ($('#stream_filters li.narrow-filter').length > 0) {
+
+        var selected_stream_id = $('#stream_filters li.narrow-filter.highlighted_stream')
```
We usually don't do a blank line right after an `if` statement before starting to declare things.

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+        // Classes must be explicitly named
+        if (highlighting_class === 'highlighted_stream') {
+            all_streams.first().addClass('highlighted_stream');
+
```
I'd remove this blank line.

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+    }
+
+    if (handled) {
+        // Not handle key event (do not move cursor)
```
This comment would be better as something like "Since we already handled the key event above, suppress the browser handling it."; as written it can be read as the opposite of its intent ;)

timabbott commented on static/js/stream_list.js:
```diff
@@ -529,29 +540,143 @@ exports.clear_and_hide_search = function () {
+exports.keydown_filter = function (e, all_streams_selector,
+ scroll_container, highlighting_class, enter_press_function) {
+    // Function for left and right sidebar
+    // Could be placed somewhere else but ui.js is already very full
```
Agreed.  I think we'll probably eventually want to move this and its helpers to some new file like `sidebar_ui.js` (whether that's the right name depends a bit on whether we think this could be reused for other keyboard UI situations as well).  I'd save that for a follow-up project, though, since this is already a large PR.

timabbott commented on static/js/activity.js:
```diff
@@ -568,30 +586,40 @@ exports.toggle_filter_displayed = function () {
 
+function keydown_user_filter(e) {
+    stream_list.keydown_filter(e, '#user_presences li.user_sidebar_entry.narrow-filter',
+     $('#user_presences'), 'highlighted_user', keydown_enter_key);
```
We'd usually line-wrap this to line up with the `e`.

timabbott commented on frontend_tests/casper_tests/03-narrow.js:
```diff
@@ -313,19 +313,55 @@ casper.waitWhileSelector('#streams_list .input-append.notdisplayed', function ()
 // There will be no race condition between these two waits because we
 // expect them to happen in parallel.
 casper.waitWhileVisible('#stream_filters [data-stream-name="Denmark"]', function () {
+    casper.test.info('Seach term entered');
```
"Search" is missing a letter.

timabbott commented on frontend_tests/casper_tests/03-narrow.js:
```diff
@@ -367,6 +404,48 @@ casper.waitForSelector('.input-append.notdisplayed', function () {
+
+casper.waitForSelector('#user_presences .highlighted_user', function () {
+    casper.test.info('Suggestion highlighting - initial situation');
+    casper.test.assertExist('#user_presences li.highlighted_user [data-name="Cordelia Lear"]', 'User Cordelia Lear is selected');
```
I'd line-wrap these; we're planning to start enforcing maximum line lengths for our JS codebase, and it'll save a bit of work then.

florianjuengermann commented:
I updated my PR and fixed the line-wrapping and whitespace issues. By the way, was it right to push a new commit or would you have preferred a modification of my current commits?

florianjuengermann commented:
Concerning the lint problem, I found out that the current configuration does not consider indentation in general; no error message is shown even for JS code like this:
```
if (a == 0) {
    b = 1;
b = 0;
            b = 1;
}
```

timabbott commented:
We'd prefer the fixes go into the original commits (if you look at Zulip's history, we have very few "fixup" commits).  It makes the history a lot more readable, since it isn't cluttered with reformatting.

Re: whitespace, OK, sounds we should look around at eslint configuration options.  Probably that belongs in its own issue.

florianjuengermann commented:
That makes sense. I have reorganized my commits to include the code-style changes from the beginning.

timabbott commented:
Nice, merged, thanks @florianjuengermann!  One note: I edited the second commit message before merging to address two things: (1) the `Fixes: #5920` should be included in the commit message so that when we merge the commits, the original issue is closed and (2) being a bit more clear about the interaction.

You can see the result here: ec3e0caeced1c370b9740a1cbfeeb0c2f1b05ee9

Huge thanks for building this @florianjuengermann!
